### PR TITLE
Refactor row reader/writer APIs

### PR DESCRIPTION
**Pull request by achille-roussel**
This PR is a MVP version of the proposal in https://github.com/segmentio/parquet-go/issues/149, demonstrating how the changes would modify the APIs.

### CODE REVIEW

File: buffer.go [hunk 1]
```diff
@@ -200,21 +201,29 @@ func (buf *Buffer) WriteRow(row Row) error {
...
+			columnIndex := value.Column()
+			buf.colbuf[columnIndex] = append(buf.colbuf[columnIndex], value)
+		}
 	}
 
 	for columnIndex, values := range buf.colbuf {
 		if _, err := buf.columns[columnIndex].WriteValues(values); err != nil {
-			return err
+			// TODO: we should rollback the writes that may have succeded or
+			// we are leaving the buffer in an undertermined state.
```
Comment (kevinburkesegment) [inline buffer.go:217]:
```suggestion
			// we are leaving the buffer in an undetermined state.
```

File: reader.go [hunk 2]
```diff
@@ -30,7 +30,7 @@ type Reader struct {
 	file     reader
 	read     reader
 	rowIndex int64
-	values   []Value
+	values   [1]Row
```
Comment (kevinburkesegment) [inline reader.go:33]: [1]Row feels very odd, why not just ditch the array?

File: reader_test.go [hunk 3]
```diff
@@ -369,14 +369,14 @@ func BenchmarkReaderReadRow(b *testing.B) {
 			}
 
 			r := parquet.NewReader(f)
-			rowbuf := make(parquet.Row, 0, 16)
+			rowbuf := make([]parquet.Row, 20)
```
Comment (kevinburkesegment) [inline reader_test.go:372]: `defaultRowBufferSize` here?

**Review by kevinburkesegment — APPROVED** [review]

**Comment by achille-roussel** [comment]
Closing in favor of https://github.com/segmentio/parquet-go/pull/183

**Pull request closed without merging**